### PR TITLE
Decoded call hash and length

### LINKED ISSUE #103: Auto-generate `hash` and `len` variables for every decoded call
It would be very handy to have a hash and a len for every decoded call automatically placed into the variables.
For example if there a `decodedCall.send_xcm_to_cp`, which can be referenced as `$send_xcm_to_cp`, an auto-generated variables will be referenced as `$__hash_send_xcm_to_cp` and `$__len_send_xcm_to_cp`.

Use case:
For referenda pallet every proposal's call has to be uploaded to the preimage pallet. The proposal's call will be referenced for the referenda pallet by its `hash` and `len`.

### PR DESCRIPTION
**Pull request by muharem**
fixes: https://github.com/paritytech/parachains-integration-tests/issues/103

### CODE REVIEW

File: src/events.ts [hunk 1]
```diff
@@ -29,12 +30,12 @@ const messageBuilder = (context, event: EventResult): string => {
...
     if (result) {
-      let resultJson = JSON.stringify(result);
+      let resultJson = JSON.stringify(parseArgs(context, result));
       let recordJson = JSON.stringify(record);
-      let isExpectedResult = isExpectedEventResult(event);
+      let isExpectedResult = isExpectedEventResult(context, event);
 
       if (isExpectedResult) {
-        resultMessage = `\n\n   ✔️  Expected Result: ${resultJson}`;
+        resultMessage = `\n\n   ✔️  Expected Result: ${resultJson} | Received Result: ${recordJson}`;
```
Comment (muharem) [inline src/events.ts]: useful when you wanna see full result to extend the expected

Comment (NachoPal) [reply src/events.ts]: Do you need this? Because I was think of a debug mode where all received events are printed.

File: src/events.ts [hunk 2]
```diff
@@ -123,7 +124,7 @@ const messageBuilder = (context, event: EventResult): string => {
 };
 
 const isStrict = (event: Event): boolean => {
-  if (event.strict == undefined && event.threshold != undefined) {
+  if (event.strict == undefined || event.threshold != undefined) {
```
Comment (NachoPal) [inline src/events.ts]: I am unsure about this. Did you check the rest of tests (assets) still properly work with this change?

File: src/before.ts [hunk 3]
```diff
@@ -61,17 +61,13 @@ export const beforeBuildDecodedCalls = (decodedCalls) => {
...
-              decodedCalls[key]
-            );
-          }
+          const { encoded, hash, len } = 
+            decodedCalls[key].encode === false ? 
+              buildEncodedCallHex(this, decodedCalls[key]) :
+              buildEncodedCall(this, decodedCalls[key]);
+          this.variables[`\$${key}`] = encoded;
+          this.variables[`\$_hash__${key}`] = hash;
+          this.variables[`\$_len__${key}`] = len;
```
Comment (NachoPal) [inline src/before.ts]: I'd like to suggest creating the new keys in this way instead, mocking an object access. We could potentially add more in the future or even modify the parser to actually access the object.

```suggestion
          this.variables[`\$${key}.hash`] = hash;
          this.variables[`\$${key}.len`] = len;
```

File: src/events.ts (hunk 2 above)
Comment (muharem) [reply src/events.ts]: otherwise `threshold` wont work

Comment (muharem) [reply src/events.ts]: if strict = true, and threshold something, threshold wont work

**Review by NachoPal — APPROVED** [review]

**Pull request merged**